_episodes/02-hardware-performance.md

```diff
@@ -15,7 +15,74 @@ usemathjax: true
+
+Similar situation arises for GPU coding too. In this case, the CPU is generally called the host and the GPUs are called the devices. When we submit a GPU job, it is launched in the CPU (host) which in turn directs it to be executed by the GPU cores (devices). While doing these calculations, data is copied from CPU memory to GPU’s first and processed subsequently. After finishing a calculation by a GPU, the result is copied back from GPU to CPU. This communication process is expensive and it could significantly slow down a calculation if care is not taken to minimize this communication. We’ll see later in this tutorial that communication is a major bottleneck in many LAMMPS calculations and we need to device strategies to reduce this communication overhead.  
+
+On the contrary, in a shared memory platforms, the central memory is being shared by all the processors. In this case, the processors communicate with each other through the shared memory. In a shared memory platform, each processor can access the memory location at the same speed. 
+
+When we say that we accelerate a job by parallelizing it, we actually mean a strategy that divide the whole job into pieces and assign each piece to a worker (CPU core) to solve. This parallelisation strategy also depends on the memory structure. (as discussed above) of your computing system. For example, OpenMP provides a thread level parallelism that is well-suited for a shared memory platform, but you can not use it in a distributed memory system. For a distributed memory system, you need a message passing protocol like MPI to communicate between processors. 
+
+The two main parallelization strategies are data parallelism and task parallelism. In data parallelism same set of tasks are performed by each core on different subsets of the same data. You need to take extra care to eliminate data dependencies in data parallelism. On the contrary, when a task is multiple independent sub-tasks and each sub-task is assigned to different cores to do the job, this is known as task parallelism. In one sentence, in task parallelism different independent sub-tasks are performed on the same or different data. A simple graphical representation is given below:
+
+[image here]
+
+
+* Process vs. Threads
+
+Just imagine that you are doing some work with both of your hands. You are a process and your both hands are the threads. In a program, threads are separate points of execution and depending on the circumstances these two threads either can work synchronously or in asynchronous manner.
+
+* Multithreading
+
+Since we mentioned about threads, its good to know that there are two main methods to introduce thread-based computation: either use OpenMP or CUDA. OpenMP is used for multi-core CPUs and CUDA is used to utlize threading for the GPU cores.
```

Comment:
There are more threading options

Reply:
Replaced with "Multithreading allows a single process to contain multiple threads which share the process's resources but execute independently. This is important since it allows a single process to execute parallel processing using many threads on multiprocessor systems. In practice, this involves many difficulties. OpenMP and CUDA are two very popular thread-level execution model of parallel processing. OpenMP is used for multi-core/manycore CPUs and CUDA is used to utlize threading for the GPU cores."